sql/sql_table.cc

```diff
@@ -775,7 +775,23 @@ bool mysql_write_frm(ALTER_PARTITION_PARAM_TYPE *lpt, uint flags)
   strxmov(shadow_frm_name, shadow_path, reg_ext, NullS);
   if (flags & WFRM_WRITE_SHADOW)
   {
-    if (mysql_prepare_create_table(lpt->thd, lpt->create_info, lpt->alter_info,
+    /*
+      It is possible mysql_prepare_create_table was already called in our
+      create/alter_info context and we don't need to call it again. That is, if
+      in the context of `ALTER TABLE ... CONVERT PARTITION .. TO TABLE` then
+      mysql_prepare_create_table would have already been called through a prior
+      invocation of mysql_write_frm with flag MFRM_WRITE_CONVERTED_TO.
+
+      We check if mysql_prepare_create_table was already called by checking if
+      the alter_info struct's keys have already been initialized, which is done
+      in mysql_prepare_create_table_finalize (via function init_key_info()).
+    */
+    bool prepare_create_table_already_called= false;
+    if (!lpt->alter_info->key_list.is_empty())
```

Comment:
So, do you think it may be fine to skip calling `mysql_prepare_create_table` if `key_list` is empty? If so, it means we need it only to initialize the keys.

Reply:
For all cases? Or just `CONVERT_OUT`? And for both `WFRM_WRITE_CONVERTED_TO` and `WFRM_WRITE_SHADOW`?

I'd still think even for `CONVERT_OUT`, it should be done for the first call to `mysql_write_frm` even when `key_list` is empty, as it does other checks and verifications (despite it being created from a definition of an existing table, I'd always err on the side of validation).